### PR TITLE
support tasks for managing kubernetes custom resources

### LINKED ISSUE #44: Support Custom Objects API
# Expectation / Proposal
Add `CustomObjectsAPI` to mange crud operations on Kubernetes Custom Resource Definitions. This gives end users flexibility to create their own operator specific resources and watch over them.

# Traceback / Example

- [x] I would like to [help contribute](https://PrefectHQ.github.io/prefect-kubernetes/#contributing) a pull request to resolve this!

### PR DESCRIPTION
**Pull request by tardunge**
Closes #44 

### Screenshots
Mkdocs:
Screenshot 2023-03-26 at 12 08 01 AM

Service Integration tests:
Screenshot 2023-03-26 at 12 20 12 AM

### Checklist

- [x] References any related issue by including "Closes #<Issue Number>" or "Closes <Issue URL>".
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect-kubernetes/issues/new/choose) first.
- [x] Includes tests or only affects documentation.
- [x] Passes `pre-commit` checks.
  - Run `pre-commit install && pre-commit run --all` locally for formatting and linting.
- [x] Includes screenshots of documentation updates.
  - Run `mkdocs serve` view documentation locally.
- [x] Summarizes PR's changes in [CHANGELOG.md](https://github.com/PrefectHQ/prefect-kubernetes/blob/main/CHANGELOG.md)

### CODE REVIEW

**Comment by tardunge** [comment]
Contains only crud operations for namespaced scope.
Should we add cluster scoped operations as well?

File: prefect_kubernetes/custom_objects.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+        kubernetes_credentials: KubernetesCredentials block
+            holding authentication needed to generate the required API client.
+        group: The custom resource object's group
+        version: The custom resource object's version
+        plural: The custom resource object's plural
+        name: The name of a custom object to patch.
+        body: A Dict containing the custom resource object's specification.
+        namespace: The Kubernetes namespace to patch the custom resources.
+        **kube_kwargs: Optional extra keyword arguments to pass to the
+            Kubernetes API (e.g. `{"pretty": "...", "dry_run": "..."}`).
```
Comment (desertaxle) [inline prefect_kubernetes/custom_objects.py:394]:
```suggestion
    Args:
        kubernetes_credentials: KubernetesCredentials block
            holding authentication needed to generate the required API client.
        group: The custom resource object's group
        version: The custom resource object's version
        plural: The custom resource object's plural
        name: The name of a custom object to replace.
        body: A Dict containing the custom resource object's specification.
        namespace: The custom resource's Kubernetes namespace.
        **kube_kwargs: Optional extra keyword arguments to pass to the
            Kubernetes API (e.g. `{"pretty": "...", "dry_run": "..."}`).
```

File: prefect_kubernetes/custom_objects.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+    """Task for reading a namespaced Kubernetes custom object.
+
+    Args:
+        kubernetes_credentials: `KubernetesCredentials` block
+            holding authentication needed to generate the required API client.
+        group: The custom resource object's group
+        version: The custom resource object's version
+        plural: The custom resource object's plural
+        name: The name of a custom object to read.
+        namespace: The Kubernetes namespace to read this custom resource is in.
```
Comment (desertaxle) [inline prefect_kubernetes/custom_objects.py:148]:
```suggestion
        namespace: The Kubernetes namespace the custom resource is in.
```

File: prefect_kubernetes/custom_objects.py [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+
+
+@task
+async def get_namespaced_custom_object(
+    kubernetes_credentials: KubernetesCredentials,
+    group: str,
+    version: str,
+    plural: str,
+    name: str,
+    namespace: Optional[str] = "default",
```
Comment (desertaxle) [inline prefect_kubernetes/custom_objects.py:136]:
```suggestion
    namespace: str = "default",
```

File: prefect_kubernetes/custom_objects.py [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+
+
+@task
+async def delete_namespaced_custom_object(
+    kubernetes_credentials: KubernetesCredentials,
+    group: str,
+    version: str,
+    plural: str,
+    name: str,
+    namespace: Optional[str] = "default",
```
Comment (desertaxle) [inline prefect_kubernetes/custom_objects.py:78]:
```suggestion
    namespace: str = "default",
```

File: prefect_kubernetes/custom_objects.py [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+    """Task for deleting a namespaced custom object.
+
+    Args:
+        kubernetes_credentials: `KubernetesCredentials` block
+            holding authentication needed to generate the required API client.
+        group: The custom resource object's group
+        version: The custom resource object's version
+        plural: The custom resource object's plural
+        name: The name of a custom object to delete.
+        namespace: The Kubernetes namespace to create this custom object in.
```
Comment (desertaxle) [inline prefect_kubernetes/custom_objects.py:90]:
```suggestion
        namespace: The Kubernetes namespace the custom object is in.
```

File: prefect_kubernetes/custom_objects.py [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+
+
+@task
+async def create_namespaced_custom_object(
+    kubernetes_credentials: KubernetesCredentials,
+    group: str,
+    version: str,
+    plural: str,
+    body: Dict[str, Any],
+    namespace: Optional[str] = "default",
```
Comment (desertaxle) [inline prefect_kubernetes/custom_objects.py:16]:
```suggestion
    namespace: str = "default",
```

File: prefect_kubernetes/custom_objects.py [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+    """Task for creating a namespaced custom object.
+
+    Args:
+        kubernetes_credentials: `KubernetesCredentials` block
+            holding authentication needed to generate the required API client.
+        group: The custom resource object's group
+        version: The custom resource object's version
+        plural: The custom resource object's plural
+        body: A Dict containing the custom resource object's specification.
+        namespace: The Kubernetes namespace to create this custom object in.
```
Comment (desertaxle) [inline prefect_kubernetes/custom_objects.py:28]:
```suggestion
        namespace: The Kubernetes namespace to create the custom object in.
```

File: prefect_kubernetes/custom_objects.py [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+
+
+@task
+async def get_namespaced_custom_object_status(
+    kubernetes_credentials: KubernetesCredentials,
+    group: str,
+    version: str,
+    plural: str,
+    name: str,
+    namespace: Optional[str] = "default",
```
Comment (desertaxle) [inline prefect_kubernetes/custom_objects.py:195]:
```suggestion
    namespace: str = "default",
```

File: prefect_kubernetes/custom_objects.py [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+        )
+
+
+@task
+async def list_namespaced_custom_object(
+    kubernetes_credentials: KubernetesCredentials,
+    group: str,
+    version: str,
+    plural: str,
+    namespace: Optional[str] = "default",
```
Comment (desertaxle) [inline prefect_kubernetes/custom_objects.py:252]:
```suggestion
    namespace: str = "default",
```

File: prefect_kubernetes/custom_objects.py [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+
+@task
+async def patch_namespaced_custom_object(
+    kubernetes_credentials: KubernetesCredentials,
+    group: str,
+    version: str,
+    plural: str,
+    name: str,
+    body: Dict[str, Any],
+    namespace: Optional[str] = "default",
```
Comment (desertaxle) [inline prefect_kubernetes/custom_objects.py:307]:
```suggestion
    namespace: str = "default",
```

File: prefect_kubernetes/custom_objects.py [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+
+@task
+async def replace_namespaced_custom_object(
+    kubernetes_credentials: KubernetesCredentials,
+    group: str,
+    version: str,
+    plural: str,
+    name: str,
+    body: Dict[str, Any],
+    namespace: Optional[str] = "default",
```
Comment (desertaxle) [inline prefect_kubernetes/custom_objects.py:379]:
```suggestion
    namespace: str = "default",
```

**Review by desertaxle — CHANGES_REQUESTED** [review]
Thanks for your contribution @tardunge! 

Overall this looks great! In addition to the comments on typings and docstrings, could you please add an entry to `CHANGELOG.md`?

**Comment by tardunge** [comment]
> Thanks for your contribution @tardunge!
> 
> Overall this looks great! In addition to the comments on typings and docstrings, could you please add an entry to `CHANGELOG.md`?

Done.

File: CHANGELOG.md [hunk 12]
```diff
@@ -19,6 +19,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Security
 
+## 0.2.4
+
+Released March 27, 2023.
+
+### Added
+- Custom objects crud tasks for kubernetes custom resource definitions. - [#45](https://github.com/PrefectHQ/prefect-kubernetes/pull/45)
```
[the remark this answers is not in the document]
Comment (tardunge) [reply CHANGELOG.md]: done.

**Review by desertaxle — APPROVED** [review]
LGTM!

**Pull request merged**